### PR TITLE
#d2l

### PR DESCRIPTION
**Pull request by jiejieyun**
《动手学深度学习》paddle版
8.1
9.1
9.2
9.3
9.4
13.9

### CODE REVIEW

**Comment by jiejieyun** [comment]
忘记#d2l 了，手动补救

**Comment by jiejieyun** [comment]
lstm.md是9.2的内容

**Comment by jiejieyun** [comment]
deep-rnn.md是9.3的内容

**Comment by jiejieyun** [comment]
bi-rnn.md是9.4的内容

**Comment by w5688414** [comment]
1.代码重复：

```
onestep_preds = net(features)

onestep_preds = net(features)
d2l.plot([time, time[tau:]],
         [x.detach().numpy(), onestep_preds.detach().numpy()], 'time',
         'x', legend=['data', '1-step preds'], xlim=[1, 1000],
         figsize=(6, 3))
```
2.Paddle不需要detach操作，直接.numpy()就行。

```
d2l.plot([time[tau + i - 1: T - max_steps + i] for i in steps],
    [features[:, (tau + i - 1)].detach().numpy() for i in steps], 'time', 'x',
    legend=[f'{i}-step preds' for i in steps], xlim=[5, 1000],
    figsize=(6, 3))
```

**Comment by jiejieyun** [comment]
已更新，请查验

**Comment by 1649759610** [comment]
@jiejieyun 同学你好，代码尚存在一些细节问题，请确认和修改，具体如下：

### 9.1 节 
Return 元组中的逗号问题

def init_gru_state(batch_size, num_hiddens, device):
    return (paddle.zeros([batch_size, num_hiddens], ))
—>
def init_gru_state(batch_size, num_hiddens, device):
    return (paddle.zeros([batch_size, num_hiddens]), )

def gru(inputs, state, params):
    W_xz, W_hz, b_z, W_xr, W_hr, b_r, W_xh, W_hh, b_h, W_hq, b_q = params
    H,*_ = state
    outputs = []
    for X in inputs:
        Z = paddle.nn.functional.sigmoid((X @ W_xz) + (H @ W_hz) + b_z)
        R = paddle.nn.functional.sigmoid((X @ W_xr) + (H @ W_hr) + b_r)
        H_tilda = paddle.tanh((X @ W_xh) + ((R * H) @ W_hh) + b_h)
        H = Z * H + (1 - Z) * H_tilda
        Y = H @ W_hq + b_q
        outputs.append(Y)
    return paddle.concat(outputs, axis=0), (H,*_)

—>可以简洁为：
import paddle.nn.functional as F
F.sigmoid((X @ W_xz) + (H @ W_hz) + b_z) 

—> return 时 带了一个*_， 与原书代码不太一致，请check，如信息后续没有使用，请删除*_

简洁实现
num_inputs = vocab_size
num_epochs, lr = 500, 1.0
rnn_layer = nn.GRU(len(vocab), num_hiddens, time_major=True)
net = d2l.RNNModel(rnn_layer, vocab_size=len(vocab))
d2l.train_ch8(net, train_iter, vocab, lr, num_epochs, device)

—>
代码块中已经定义num_inputs， len(vocab)可统一换成num_inputs, 另外num_epochs, lr上个代码块已经声明，这里无需二次声明
代码运行中有连续输出一些shape，请检查删除该print信息


### 9.2节

def init_lstm_state(batch_size, num_hiddens, device):
    return (paddle.zeros([batch_size, num_hiddens]),
            paddle.zeros([batch_size, num_hiddens]))
—> 函数中没有用到device，可以删除

Lstm的模型定义： paddle.nn.functional —> F.sigmoid


简介实现
num_inputs = vocab_size
num_epochs, lr = 500, 1.0
lstm_layer = nn.LSTM(num_inputs, num_hiddens, time_major=True)
model = d2l.RNNModel(lstm_layer, len(vocab))
d2l.train_ch8(model, train_iter, vocab, lr, num_epochs, device)

num_epochs, lr 上个代码块中已经定义，对齐原书，这里考虑删除

### 9.3节

import paddle
from paddle import nn
from d2l import paddle as d2l

batch_size, num_steps = 32, 32
train_iter, vocab = d2l.load_data_time_machine(batch_size, num_steps)


—>原书中设置num_steps=35


### 9.4节

import paddle
from paddle import nn
from d2l import paddle as d2l

#加载数据
batch_size, num_steps, device  = 32, 35, d2l.try_gpu()
train_iter, vocab = d2l.load_data_time_machine(batch_size, num_steps)
#通过设置“bidirective=True”来定义双向LSTM模型
vocab_size, num_hiddens, num_layers = len(vocab), 256, 2
num_inputs = vocab_size
lstm_layer = nn.LSTM(num_inputs, num_hiddens, num_layers, direction='bidirect', time_major=True)
model = d2l.RNNModel(lstm_layer, len(vocab))
#训练模型
num_epochs, lr = 500, 1.0
d2l.train_ch8(model, train_iter, vocab, lr, num_epochs, device)


注释错误：#通过设置“bidirective=True”来定义双向LSTM模型 —> 设置direction=“”bidirect

训练过程中打印信息：time travellerererererererererererererererererererererererererer，请检查是否打印有误

**Comment by jiejieyun** [comment]
@1649759610 请大佬确认是否还需要修改，若不需要我就更新一下，若还需要修改请告知
**9.1 节**
Return 元组中的逗号问题

def init_gru_state(batch_size, num_hiddens, device):
return (paddle.zeros([batch_size, num_hiddens], ))
—>**#已修改**
def init_gru_state(batch_size, num_hiddens, device):
return (paddle.zeros([batch_size, num_hiddens]), )

def gru(inputs, state, params):
W_xz, W_hz, b_z, W_xr, W_hr, b_r, W_xh, W_hh, b_h, W_hq, b_q = params
H,_ = state
outputs = []
for X in inputs:
Z = paddle.nn.functional.sigmoid((X @ W_xz) + (H @ W_hz) + b_z)
R = paddle.nn.functional.sigmoid((X @ W_xr) + (H @ W_hr) + b_r)
H_tilda = paddle.tanh((X @ W_xh) + ((R * H) @ W_hh) + b_h)
H = Z * H + (1 - Z) * H_tilda
Y = H @ W_hq + b_q
outputs.append(Y)
return paddle.concat(outputs, axis=0), (H,_)

—>**#已修改**
可以简洁为：
import paddle.nn.functional as F
F.sigmoid((X @ W_xz) + (H @ W_hz) + b_z)

—> return 时 带了一个*， 与原书代码不太一致，请check，如信息后续没有使用，请删除*
**#这里使用到了该信息，若删除*会产生错误**

简洁实现
num_inputs = vocab_size
num_epochs, lr = 500, 1.0
rnn_layer = nn.GRU(len(vocab), num_hiddens, time_major=True)
net = d2l.RNNModel(rnn_layer, vocab_size=len(vocab))
d2l.train_ch8(net, train_iter, vocab, lr, num_epochs, device)

—>
代码块中已经定义num_inputs， len(vocab)可统一换成num_inputs, 另外num_epochs, lr上个代码块已经声明，这里无需二次声明
代码运行中有连续输出一些shape，请检查删除该print信息
**#第一个问题已修改
#num_epochs, lr = 500, 1.0由于paddle的部分问题导致输入lr = 1会产生错误learning rate should be float or LRScheduler, got <class 'int'> here，这里需输入1.0，故增加了该行**

**9.2节**
def init_lstm_state(batch_size, num_hiddens, device):
return (paddle.zeros([batch_size, num_hiddens]),
paddle.zeros([batch_size, num_hiddens]))
—> 函数中没有用到device，可以删除
**#此处未用的device，但是d2l包内输入需要device若删除则报以下错误init_lstm_state() takes 2 positional arguments but 3 were given，故未修改**

Lstm的模型定义： paddle.nn.functional —> F.sigmoid
**#此处由于存在变量F，若修改成F.sigmoid则报以下错误local variable 'F' referenced before assignment，故将
paddle.nn.functional修改为Function.sigmoid**

简介实现
num_inputs = vocab_size
num_epochs, lr = 500, 1.0
lstm_layer = nn.LSTM(num_inputs, num_hiddens, time_major=True)
model = d2l.RNNModel(lstm_layer, len(vocab))
d2l.train_ch8(model, train_iter, vocab, lr, num_epochs, device)

num_epochs, lr 上个代码块中已经定义，对齐原书，这里考虑删除
**#num_epochs, lr = 500, 1.0由于paddle的部分问题导致输入lr = 1会产生错误learning rate should be float or LRScheduler, got <class 'int'> here，这里需输入1.0，故增加了该行****

**9.3节**
import paddle
from paddle import nn
from d2l import paddle as d2l

batch_size, num_steps = 32, 32
train_iter, vocab = d2l.load_data_time_machine(batch_size, num_steps)

—>原书中设置num_steps=35
**#已修改**

9.4节
import paddle
from paddle import nn
from d2l import paddle as d2l

#加载数据
batch_size, num_steps, device = 32, 35, d2l.try_gpu()
train_iter, vocab = d2l.load_data_time_machine(batch_size, num_steps)
#通过设置“bidirective=True”来定义双向LSTM模型
vocab_size, num_hiddens, num_layers = len(vocab), 256, 2
num_inputs = vocab_size
lstm_layer = nn.LSTM(num_inputs, num_hiddens, num_layers, direction='bidirect', time_major=True)
model = d2l.RNNModel(lstm_layer, len(vocab))
#训练模型
num_epochs, lr = 500, 1.0
d2l.train_ch8(model, train_iter, vocab, lr, num_epochs, device)

注释错误：#通过设置“bidirective=True”来定义双向LSTM模型 —> 设置direction=“”bidirect
**#已修改**

训练过程中打印信息：time travellerererererererererererererererererererererererererer，请检查是否打印有误
**#最终结果与书本一致
困惑度 1.1, 126711.0 词元/秒 CUDAPlace(0)
time travellerererererererererererererererererererererererererer
travellerererererererererererererererererererererererererer
这里训练过程中打印该信息应该是正常的**

**Comment by 1649759610** [comment]
@jiejieyun 辛苦同学及时修改：

9.1，9.2节 
可以考虑：在【训练与预测】小节中，将set lr=1.0 ， 然后在简洁实现不重复。学习率本身即应该为float类型，因此声明时最好直接set为float类型。


以下问题请彤彤  @tngt 确认d2l包的相关函数设置是否合理，如不合理，可考虑修改一下d2l包中函数。
#此处未用的device，但是d2l包内输入需要device若删除则报以下错误init_lstm_state() takes 2 positional arguments but 3 were given，故未修改
以上问题在get_lstm_params也存在。



其他无问题，辛苦 @jiejieyun  @tngt 。

**Comment by jiejieyun** [comment]
@1649759610 @tngt
其他问题均已解决。
关于device的问题在删除device后对d2l包相关函数修改后可以跑通一步一步实现的情况，但是由于简洁实现使用了相同的函数（train_epoch_ch8）就会产生错误。所以不建议删除device，若一定要删除可以向d2l包重新再写几组类似的函数来单独实现简洁实现，但是这样d2l包会冗余。

**Comment by jiejieyun** [comment]
@1649759610 @tngt
device的问题也解决了，需要修改d2l包函数，我会向相关同学反映
请及时查验

**Comment by 1649759610** [comment]
review 无问题 @jiejieyun @tngt

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+           + colormap[:, :, 2])
+    return colormap2label[idx]
+```
+
+[**例如**]，在第一张样本图像中，飞机头部区域的类别索引为1，而背景索引为0。
+
+
+
+```python
+y = voc_label_indices(train_labels[0], voc_colormap2label())
```
Comment (ZhangHandi) [inline Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:143]: 这里aistudio结果为 (Tensor(shape=[10, 10], dtype=int64, place=CUDAPlace(0), stop_gradient=True,
        [[0, 0, 0, 0, 0, 0, 0, 0, 0, 4],
         [0, 0, 0, 0, 0, 0, 0, 4, 4, 4],
         [0, 0, 0, 0, 0, 0, 4, 4, 4, 4],
         [0, 0, 0, 0, 0, 4, 4, 4, 4, 4],
         [0, 0, 0, 0, 0, 4, 4, 4, 4, 4],
         [0, 0, 0, 0, 4, 4, 4, 4, 4, 4],
         [0, 0, 0, 0, 0, 4, 4, 4, 4, 4],
         [0, 0, 0, 0, 0, 4, 4, 4, 4, 4],
         [0, 0, 0, 0, 0, 0, 4, 4, 4, 4],
         [0, 0, 0, 0, 0, 0, 0, 0, 4, 4]]),
 'aeroplane')
和书中不同

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    Defined in :numref:`sec_semantic_segmentation`"""
+    rect = paddle.vision.transforms.RandomCrop((height, width))._get_param(
+        img=feature, output_size=(height, width))
+    feature = paddle.vision.transforms.crop(feature, *rect)
+    label = paddle.vision.transforms.crop(label, *rect)
+    return feature, label
+```
+
+
+```python
```
Comment (ZhangHandi) [inline Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:170]: 和上一个代码块 voc_rand_crop 合并

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+设批量大小为64，我们定义训练集的迭代器。
+打印第一个小批量的形状会发现：与图像分类或目标检测不同，这里的标签是一个三维数组。
+
+
+
+```python
+batch_size, num_workers = 64, 0
+train_iter = paddle.io.DataLoader(voc_train, batch_size=batch_size, shuffle=True,
+                                    drop_last=True,
+                                    num_workers=num_workers)
```
Comment (ZhangHandi) [inline Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:246]: 这里书中是 num_workers=d2l.get_dataloader_workers()，为什么自己定义 num_workers 呢

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+print(train_iter)
```
Comment (ZhangHandi) [inline Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:247]: 这一行删掉

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+for X, Y in  train_iter:
```
Comment (ZhangHandi) [inline Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:248]: 这里in后面多空了一个格

**Comment by jiejieyun** [comment]
@ZhangHandi 
**第一个问题：** 由于我是在aistdio上运行，完整运行完需要三次运行到
train_features, train_labels = read_voc_images(voc_dir, True)
故需要大约40G显存（至尊版仅有31.7G），所以将train_features, train_labels = read_voc_images(voc_dir, True)与上一块代码分开，这样运行后面半块代码重启执行器即可（这样就能少执行一次）。
理论上显存足够大就没问题。若确认后需要合并就合并。

**第二个问题：** 该pr提交过早未及时更新，在aistdio上输出与原书一致。确认后会立即更新。

**第三个问题：** 同问题二。

**第四个问题：** 这个问题是因为aistdio上numworks只能设置为0，而书中num_workers=d2l.get_dataloader_workers()返回的是4，会产生错误，该问题在群里已经讨论过了，并最终用这种方法解决。

**第五个问题：** 已修改。

**第六个问题：** 已修改。

**请及时查验，若无问题我就更新一下pr。**

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        features.append(paddle.to_tensor(paddle.vision.image.image_load(os.path.join(
+            voc_dir, 'JPEGImages', f'{fname}.jpg'), backend='cv2'), dtype=paddle.float32).transpose([2,0,1]))
+        labels.append(paddle.to_tensor(paddle.vision.image.image_load(os.path.join(
+            voc_dir, 'SegmentationClass' ,f'{fname}.png'), backend='cv2'), dtype=paddle.float32).transpose([2,0,1]))
+    return features, labels
+```
+
+
+```python
+train_features, train_labels = read_voc_images(voc_dir, True)
```
[the remark this answers is not in the document]
Comment (ZhangHandi) [reply Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:79]: 这个代码块要合并

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md (hunk 3 above)
Comment (ZhangHandi) [reply Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:240]: 这里为什么是指定num_workers呢

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiejieyun) [reply Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:76]: 已合并

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md (hunk 3 above)
Comment (jiejieyun) [reply Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:237]: def get_dataloader_workers():
    """使用4个进程来读取数据。"""
    return 4

因为这个函数最新的d2l包内这个函数返回的还是4没有更改。

Comment (jiejieyun) [reply Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md]: 已修改

**Comment by ZhangHandi** [comment]
> @ZhangHandi **第一个问题：** 由于我是在aistdio上运行，完整运行完需要三次运行到 train_features, train_labels = read_voc_images(voc_dir, True) 故需要大约40G显存（至尊版仅有31.7G），所以将train_features, train_labels = read_voc_images(voc_dir, True)与上一块代码分开，这样运行后面半块代码重启执行器即可（这样就能少执行一次）。 理论上显存足够大就没问题。若确认后需要合并就合并。
> 
> **第二个问题：** 该pr提交过早未及时更新，在aistdio上输出与原书一致。确认后会立即更新。
> 
> **第三个问题：** 同问题二。
> 
> **第四个问题：** 这个问题是因为aistdio上numworks只能设置为0，而书中num_workers=d2l.get_dataloader_workers()返回的是4，会产生错误，该问题在群里已经讨论过了，并最终用这种方法解决。
> 
> **第五个问题：** 已修改。
> 
> **第六个问题：** 已修改。
> 
> **请及时查验，若无问题我就更新一下pr。**

好的，抱歉在二次review时没有看到这里的comment，第一个问题和第四个问题可以保留不做修改，第四个问题这里如果在aistudio上指定numworkers=0，可以加一段注释解释一下原因，我在整体运行aistudio的时候会出现out of memory 的问题，这是什么原因呢

**Comment by jiejieyun** [comment]
@ZhangHandi 
整体运行out of memory是因为数据集太大，他一共执行read_voc_images(voc_dir, is_train=True)三次，每次分别为train_features, train_labels/voc_train /voc_test ，所以整体运行就会显存不够。这也就是第一个问题的原因，但是理论上显存足够大，就不会出现这样的问题。

**Comment by ZhangHandi** [comment]
> @ZhangHandi
> 整体运行out of memory是因为数据集太大，他一共执行read_voc_images(voc_dir, is_train=True)三次，每次分别为train_features, train_labels/voc_train /voc_test ，所以整体运行就会显存不够。这也就是第一个问题的原因，但是理论上显存足够大，就不会出现这样的问题。

这里在每次out of memory的地方加一段话解释一下吧，解释一下整体运行显存会不够，需要重启环境再跑后续部分之类的，需要确保用户在跑aistudio跑不通时能有跑通的解决方案

**Comment by jiejieyun** [comment]
@ZhangHandi 已添加注释

File: Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md [hunk 7]
```diff
@@ -224,6 +224,7 @@ class VOCSegDataset(paddle.io.Dataset):
 
 
 ```python
+#由于数据集过大，整体运行会超显存，请在运行到这一步的时候重启内核并将所有带#save的代码段重新运行一遍
```
Comment (ZhangHandi) [inline Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md:227]: 这里我看有的代码块是#@save 有的是#save 是不是要明确一下究竟是哪种呢

Comment (jiejieyun) [reply Dive-into-DL-paddlepaddle/docs/13_computer-vision/13.9 semantic-segmentation-and-dataset.md]: 已修改

**Comment by ZhangHandi** [comment]
13.9 LGTM ai studio 记得同步更新下

**Comment by jiejieyun** [comment]
已同步

**Comment by tngt** [comment]
review全部通过可以合入

**Pull request merged**